tensorflow/python/ops/special_math_ops.py

```diff
@@ -324,13 +324,43 @@ def sort_key(input_index, a):
     num_summed_elements = _total_size(t0_shape[-len(axes_to_sum):])
     new_shape = t0_shape[:len(preserved_axes)] + (num_broadcast_elements_t0,
                                                   num_summed_elements)
+    # Make new_shape valid, if invalid to be used in reshape
+    none_dims = [i for i, d in enumerate(new_shape)
```

Comment:
Rather than defining new_shape in terms of get_shape() values (on line 325) and then fixing it here if necessary, could we just define new_shape in terms of array_ops.shape initially?  In other words, something like:

t0_shape = array_ops.shape(t0)
t1_shape = array_ops.shape(t1)
num_summed_elements = math_ops.reduce_prod(t0_shape[len(preserved_axes):-len(axes_to_sum)])
num_broadcast_elements_t1 = math.ops.reduce_prod(t0_shape[-len(axes_to_sum):])
new_shape = t1_shape[:len(preserved_axes)] + (num_summed_elements, num_broadcast_elements_t1)

You could do the same for lines 344-348, and then remove the _total_size() method.  It seems like this would also let you remove the logic for fixing uncompacted_shape below.